stdlib/dynarray.ml

```diff
@@ -0,0 +1,727 @@
+    *)
+    if n <= 512 then n * 2
+    else n + n / 2
+  in
+  (* jump directly from 0 to 8 *)
+  min (max 8 n') Sys.max_array_length
+
+let ensure_capacity a capacity_request =
+  let arr = a.arr in
+  let cur_capacity = Array.length arr in
+  if cur_capacity >= capacity_request then
+    (* This is the fast path, the code up to here must do as little as
+       possible. (This is why we don't use [let {arr; length} = a] as
+       usual, the length is not needed in the fast path.)*)
+    ()
+  else begin
+    if capacity_request < 0 then
+      Error.negative_capacity "ensure_capacity" capacity_request;
+    if capacity_request > Sys.max_array_length then
+      Error.requested_length_out_of_bounds "ensure_capacity" capacity_request;
+    let new_capacity =
+      (* We use either the next exponential-growth strategy,
+         or the requested strategy, whichever is bigger.
+
+         Compared to only using the exponential-growth strategy, this
+         lets us use less memory by avoiding any overshoot whenever
+         the capacity request is noticeably larger than the current
+         capacity.
+
+         Compared to only using the requested capacity, this avoids
+         losing the amortized guarantee: we allocated "exponentially
+         or more", so the amortization holds. In particular, notice
+         that repeated calls to [ensure_capacity a (length a + 1)]
+         will have amortized-linear rather than quadratic complexity.
+      *)
+      max (next_capacity cur_capacity) capacity_request in
+    let new_arr = Array.make new_capacity Empty in
+    Array.blit arr 0 new_arr 0 a.length;
+    a.arr <- new_arr;
+    assert (0 <= capacity_request);
+    assert (capacity_request <= Array.length new_arr);
+  end
+
+let ensure_extra_capacity a extra_capacity_request =
+  ensure_capacity a (length a + extra_capacity_request)
+
+let fit_capacity a =
+  if capacity a = a.length
+  then ()
+  else a.arr <- Array.sub a.arr 0 a.length
+
+let truncate_capacity a n =
+  if n >= capacity a then ()
+  else if n < 0 then
+    Error.negative_capacity "truncate_capacity" n
+  else begin
+    a.length <- n;
+    a.arr <- Array.sub a.arr 0 n;
+  end
+
+let reset a =
+  truncate_capacity a 0
+
+(** {1:adding Adding elements} *)
+
+(* We chose an implementation of [add_last a x] that behaves correctly
+   in presence of aynchronous code execution around allocations and
+   poll points: if another thread or a callback gets executed on
+   allocation, we add the element at the new end of the dynamic array.
+
+   (We do not give the same guarantees in presence of concurrent
+   updates, which are much more expansive to protect against.)
+*)
+
+(* [add_last_if_room a elem] only writes the slot if there is room, and
+   returns [false] otherwise.
+
+   It is sequentially atomic -- in absence of unsychronized concurrent
+   uses, the fields of [a.arr] and [a.length] will not be mutated
+   by any other code during execution of this function.
+*)
```

Comment:
I added a bit more explanations in comment. (I don't want to insist too much because this is outside the functional specification anyway: we have slightly stronger guarantees but it's no big deal.)